### PR TITLE
Pass context to all remote.Client operations

### PR DESCRIPTION
**Pull request by marcinwyszynski**
Given that many state backend methods use contexts in their calls, we will make the related interfaces accept context as an input to all methods. Since it's a ton of work, we go slowly. In this PR state reading and writing (as they're heavily intertwined) are taught to use contexts.

related to https://github.com/opentofu/opentofu/issues/753

## Target Release

1.6.0

## Draft CHANGELOG entry

n/a, it's a purely internal change

### CODE REVIEW

File: internal/backend/local/backend.go [hunk 1]
```diff
@@ -350,7 +350,7 @@ func (b *Local) opWait(
 
 		// try to force a PersistState just in case the process is terminated
 		// before we can complete.
-		if err := opStateMgr.PersistState(nil); err != nil {
+		if err := opStateMgr.PersistState(cancelCtx, nil); err != nil {
```
Comment (marcinwyszynski) [inline internal/backend/local/backend.go:353]: This is one that I'm not sure about. Perhaps we should actually use a totally separate context here?

Comment (mbialon) [reply internal/backend/local/backend.go:353]: We should. Canceling that operation was not possible before. It looks like the last line of defense against losing state changes. Canceling it midway could have bad consequences, right?

Comment (tomasmik) [reply internal/backend/local/backend.go:353]: I would either go with a separate context or maybe `stopCtx`? Though separate sounds like a better choice

**Review by mbialon — APPROVED** [review]

**Review by tomasmik — APPROVED** [review]

File: internal/backend/local/backend.go (hunk 1 above)
Comment (marcinwyszynski) [reply internal/backend/local/backend.go]: I created a separate context and documented this choice.

**Review by tomasmik — APPROVED** [review]

**Pull request merged**